packages/react/src/hooks/program.ts

```diff
@@ -0,0 +1,200 @@
+
+      if (status?.err) {
+        throw new Error(`Transaction failed: ${JSON.stringify(status.err)}`);
+      }
+
+      if (status?.confirmationStatus) {
+        const currentLevel = status.confirmationStatus;
+
+        if (currentLevel === commitment || (commitment === "confirmed" && currentLevel === "finalized")) {
+          return;
+        }
+      }
+
+      await new Promise((resolve) => setTimeout(resolve, delay));
+      delay = Math.min(delay * 1.2, CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY);
+      retryCount++;
+    } catch (error) {
+      await new Promise((resolve) => setTimeout(resolve, delay));
+      delay = Math.min(delay * 1.2, CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY);
+      retryCount++;
+      if (error instanceof Error && error.message.includes("Transaction failed")) {
+        throw error;
+      }
+    }
+  }
+}
+
+export function createProgramHook<
+  TInstructions extends Record<string, InstructionBuilder<any>>,
+  TAccounts extends Record<string, AccountFetcher<any>>,
+>(config: ProgramHookConfig<TInstructions, TAccounts>) {
+  const defaultCommitment = config.commitment === undefined ? "confirmed" : config.commitment;
+
+  function useProgramMutation<TInstructionName extends keyof TInstructions>(
+    input: UseProgramMutationInput<TInstructions, TInstructionName>,
+  ) {
+    const queryClient = useQueryClient();
+    const { instruction, commitment: mutationCommitment, ...options } = input;
+    const instructionFn = config.instructions[instruction];
+
+    return useMutation({
+      ...options,
+      mutationFn: async (input: {
+        commitment?: CommitmentLevel;
+        params: Parameters<TInstructions[TInstructionName]>[0];
+        rpc: SolanaClient['rpc'];
+        signAndSend: SignAndSendFn;
+        signer: TransactionSendingSigner;
+      }) => {
+        if (!input.params) {
+          throw new Error("Instruction params are required");
+        }
+        const { params, signer, commitment: inputCommitment, rpc, signAndSend } = input;
+        const commitmentToUse = inputCommitment ?? mutationCommitment ?? defaultCommitment;
+
+        const instruction = instructionFn(params);
+
+        const { simulateTransaction } = createSolanaClient({
+          urlOrMoniker: "devnet",
+        });
+
+        const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();
+
+        const transaction = createTransaction({
+          feePayer: signer,
+          instructions: [instruction],
+          latestBlockhash,
+          version: 0,
+        });
+
+        const simulation = await simulateTransaction(transaction);
+        console.log("simulation", simulation);
+
+        const signature = await signAndSend(instruction, signer);
+
+        if (commitmentToUse) {
+          await waitForConfirmation(rpc, signature as Signature, commitmentToUse);
+        }
+
+        await queryClient.refetchQueries({
+          queryKey: [GILL_HOOK_CLIENT_KEY, config.programAddress],
```

Comment:
I used programAddress so differentiate among multiple hooks created. There should only be one hook per program. Even if there are multiple, this allows refreshing all of them.